pkg/util/tracing/tracer.go

```diff
@@ -1092,28 +1092,28 @@ func StartVerboseTrace(ctx context.Context, tr *Tracer, opName string) (context.
 	return ctx, sp
 }
 
-// ContextWithRecordingSpan returns a context with an embedded trace Span.
-// The Span is derived from the provided Tracer. The Span returns its contents
-// when `getRecording` is called, and must be stopped using `cancel`, when done
-// with the context (`getRecording` needs to be called before `cancel`).
+// ContextWithRecordingSpan returns a context with an embedded trace Span. The
+// Span is derived from the provided Tracer. The recording is collected and the
+// span is Finish()ed through the returned callback.
+//
+// The returned callback can be called multiple times.
 //
 // Note that to convert the recorded spans into text, you can use
 // Recording.String(). Tests can also use FindMsgInRecording().
 func ContextWithRecordingSpan(
 	ctx context.Context, tr *Tracer, opName string,
-) (_ context.Context, getRecording func() Recording, cancel func()) {
+) (_ context.Context, finishAndGetRecording func() Recording) {
```

Comment:
Staring at this, can't we get to the following?

```go
  ctx, collect, finish := ContextWithRecordingSpan(...)
  defer finish()
  ...
  _ = collect()  // return recording so far
  finish()
  _ = collect() // returns stashed recording from when span was finished
  ...
```
It would get rid of the prescribed order for calling the two callbacks, which yea, was not nice. It lets us avoid the kind of overload we have elsewhere -- unclear why we're deferring a `getRecording()` and discarding the result (maybe the variables should be replaced with the more verbose `finishAndGetRecording` form? Or just be called `finish`?).

```go
recCtx, getRecording := tracing.ContextWithRecordingSpan(ctx, tr, "test")
defer getRecording()
```

I guess we're just combining the two things inline. I'd be happier if all the variables of the callback were just called `finish`.